### ISSUE TITLE
[Bug]: Slow Restores on Windows in GitHub Actions - Signature verification taking a long time

### ISSUE THREAD

**Issue by NickCraver** [NuGet/Home#11548]
### NuGet Product Used

dotnet.exe

### Product Version

6.0.101

### Repro Steps & Context

Running an identical restore in GitHub Actions, times are dramatically higher on Windows agents (2019 and 2022 tested) vs. Ubuntu agents. Here's an example run of [a `5m 39s` restore on Windows](https://github.com/StackExchange/StackExchange.Redis/runs/5029970749?check_suite_focus=true#step:4:1) vs [a `16s` restore on Ubuntu](https://github.com/StackExchange/StackExchange.Redis/runs/5029970749?check_suite_focus=true#step:4:1). It's only restoring [a hand full of packages](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/StackExchange.Redis.csproj#L17-L29).

[Here's an example PR testing various bits of this](https://github.com/StackExchange/StackExchange.Redis/pull/1973) so you can see raw logs.

Important: [setting `NUGET_CERT_REVOCATION_MODE` to `offline`](https://github.com/StackExchange/StackExchange.Redis/pull/1973/commits/275f5756cbc526df7e3bff8a82dac6dd52f2186d) indeed [improves restore times substantially](https://github.com/StackExchange/StackExchange.Redis/runs/5030088288?check_suite_focus=true#step:3:1) (down to `1m 22s` in this case).

The verbose logs [can be seen in this build](https://github.com/StackExchange/StackExchange.Redis/runs/5030159538?check_suite_focus=true). Also attaching here: [Verbose Windows Logs.txt](https://github.com/NuGet/Home/files/7982862/Verbose.Windows.Logs.txt). Scroll to `2022-02-02T00:28:54.6709200Z` for where the biggest time expenditure starts happening around package verification.

Side note: I lost the body of this issue trying to paste then directly previously (they're ~12.7MB for the restore, and the GitHub form limit is 65,536 - I recommend we don't advise people pasting logs here).

**Comment by nkolev92** [comment]
Thanks for the details @NickCraver! 

I've tweaked the title to include the verification part, to make this issue more discoverable for relevant keyword searches

**Comment by nkolev92** [comment]
This is detectable even by the basic perf scripts https://github.com/NuGet/NuGet.Client/tree/dev/scripts/perftests on my local machine.
[OrchardCore-git.csv](https://github.com/NuGet/Home/files/7989827/OrchardCore-git.csv)

Average drops from 29s -> 27s for the full restore, about 7% improvement. 

Linux vs windows probably an unfair comparison for now, but the revocation change seem to have a huge impact on windows.

**Comment by zivkan** [comment]
From Team Triage Meeting: certificate revocation checks are a security feature to ensure that a compromised certificate was not used to impersonate a trusted entity. Package signature verification has been disabled on Linux and Mac, which explains some of the perf differences. Given the design of signed packages and X509 certificates, we don't see what actionable work we can do.

**Comment by NickCraver** [comment]
@zivkan This feature takes a restore from 1 minute 22 seconds to 5+ minutes and that's not a concern at all? Forget the fact it costs me literal dollars in build minutes and that really adds up at scale, it's a security issue because _it's so bad_ my only performant option is to disable it.

There needs to be performance work invested here, or many people are going to do the same.

**Comment by aortiz-msft** [comment]
Reopening for feedback (upvotes) from the community.

**Comment by AraHaan** [comment]
I agree, when it comes to me on github actions (even though I get free minutes for open source) I expect my code to not take over 5 minutes to restore and then wait even longer for the possible chance a code change of mine might not compile. In that case I want to know the failure ASAP so it can be corrected. When it comes to CI's I could care less about security and more about things wasting my time and giving less time for me to work on more important things.

This is why I default my actions to ubuntu unless my code utilizes Windows Desktop stuff where I am forced to have wasted time.

**Comment by loic-sharma** [comment]
FYI for client team, [@bartonjs suggested](https://github.com/dotnet/runtime/issues/64689#issuecomment-1059614050) investigating this by [enabling Windows' chain-building telemetry](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-vista/cc749296(v=ws.10)?redirectedfrom=MSDN#in-this-guide).

**Comment by erdembayar** [comment]
Related: https://github.com/NuGet/Home/issues/11099

**Comment by joelverhagen** [comment]
I thought I had run into this on my project https://github.com/joelverhagen/MiniZip, but this environment variable `NUGET_CERT_REVOCATION_MODE=offline` does not seem to have a significant impact on Windows performance from my tests.

![image](https://user-images.githubusercontent.com/94054/167347573-85fcc754-40f0-486f-ac52-5848ed835e65.png)

That being said, Windows restore does appear to be significantly slower for some reason. I tried to use the setting in @NickCraver's original post but maybe I messed something up: https://github.com/joelverhagen/MiniZip/blob/338a917f132603803ccd6a90101a774ca21f40b8/.github/workflows/build.yml#L15.

I made a gist for the program that collects these timings, in case it is helpful to others. https://gist.github.com/joelverhagen/dd9123a15716863bcfed21b2b30cb8d1
